src/MLJXGBoostInterface.jl

```diff
@@ -1,770 +1,195 @@
-            "reg:"*model.objective : model.objective
-
-    seed =
-        model.seed == -1 ? generate_seed() : model.seed
-
-
-    fitresult = XGBoost.xgboost(dm, model.num_round;
-                                kwargs(model, silent, seed, objective,model.eval_metric)...)
-
-    features = schema(X).names
-    importances = [named_importance(fi, features) for
-                   fi in XGBoost.importance(fitresult)]
-    cache = nothing
-
-    report = (feature_importances=importances, )
-
-    return fitresult, cache, report
-
 end
 
+validate_reg_objective(obj) = _validate_objective(obj, ("reg:", "survival:"))
+validate_count_objective(obj) = _validate_objective(obj, ("count:", "rank:"))
+validate_class_objective(obj) = _validate_objective(obj, ("binary:", "multi:"))
+
+
+function modelexpr(name::Symbol, absname::Symbol, obj::AbstractString, objvalidate::Symbol)
+    quote
+        MMI.@mlj_model mutable struct $name <: $absname
+            test::Int = 1::(_ ≥ 0)
+            num_round::Int = 100::(_ ≥ 0)  # this must be provided since XGBoost.jl doesn't provide it
+            booster::String = "gbtree"
+            disable_default_eval_metric::Union{Bool,Int} = 0
+            eta::Float64 = 0.3::(0.0 ≤ _ ≤ 1.0)
+            num_parallel_tree::Int = 1::(_ > 0)
+            gamma::Float64 = 0::(_ ≥ 0)
+            max_depth::Int = 6::(_ ≥ 0)
+            min_child_weight::Float64 = 1::(_ ≥ 0)
+            max_delta_step::Float64 = 0::(_ ≥ 0)
+            subsample::Float64 = 1::(0 < _ ≤ 1)
+            colsample_bytree::Float64 = 1::(0 < _ ≤ 1)
+            colsample_bylevel::Float64 = 1::(0 < _ ≤ 1)
+            colsample_bynode::Float64 = 1::(0 < _ ≤ 1)
+            lambda::Float64 = 1::(_ ≥ 0)
+            alpha::Float64 = 0::(_ ≥ 0)
+            tree_method::String = "auto"
+            sketch_eps::Float64 = 0.03::(0 < _ < 1)
+            scale_pos_weight::Float64 = 1
+            updater::Union{Nothing,String} = nothing  # this is more complicated and we don't want to set it
+            refresh_leaf::Union{Int,Bool} = 1
+            process_type::String = "default"
+            grow_policy::String = "depthwise"
+            max_leaves::Int = 0::(_ ≥ 0)
+            max_bin::Int = 256::(_ ≥ 0)
+            predictor::String = "cpu_predictor"
+            sample_type::String = "uniform"
+            normalize_type::String = "tree"
+            rate_drop::Float64 = 0::(0 ≤ _ ≤ 1)
+            one_drop::Union{Int,Bool} = 0::(0 ≤ _ ≤ 1)
+            skip_drop::Float64 = 0::(0 ≤ _ ≤ 1)
+            feature_selector::String = "cyclic"
+            top_k::Int = 0::(_ ≥ 0)
+            tweedie_variance_power::Float64 = 1.5::(1 < _ < 2)
+            objective = $obj :: $objvalidate(_)
+            base_score::Float64 = 0.5
+            watchlist = nothing  # if this is nothing we will not pass it so as to use default
+            nthread::Int = Base.Threads.nthreads()::(_ ≥ 0)
+            importance_type::Union{Nothing,String} = "gain"
```

Comment:
```suggestion
            importance_type::String = "gain"
```
I believe we can do without `nothing`. Right?

Reply:
I wanted to allow for the option of eliding computation of the importance but I now realize that the way I have done this probably isn't compatible with MMI because it says it has to return an iterator of pairs.  To be honest I'm not super confident I can get the names out in the general case.  It would be nice if MMI supported being able to omit this field, so that it does not require any extra work to elide importance.

@ablaom any comments on this?

Reply:
Oh, nvm, you did it for me!!  Thanks!!